Avro parquet filesource null schema

sheivin opened the pull request.

Commit:
Read avro and parquet without setting schema

albertshau commented on core-plugins/src/main/java/co/cask/hydrator/plugin/batch/source/AbstractFileBatchSource.java:
```diff
@@ -96,7 +96,11 @@ public void configurePipeline(PipelineConfigurer pipelineConfigurer) {
+    if (config.getSchema() != null) {
+      pipelineConfigurer.getStageConfigurer().setOutputSchema(config.getSchema());
+    } else {
+      pipelineConfigurer.getStageConfigurer().setOutputSchema(DEFAULT_SCHEMA);
```
This isn't correct. For example, if schema is null and format is avro, you will set the output schema to the default schema, which is almost certainly wrong.

You should be setting the actual output schema here. The same schema you set in the input format is what should be set here.

sheivin replied:
@albertshau For setting the actual output schema in the case of when the schema is not defined, we will only be able to get the schema once the record reader has been initialized and we start reading

albertshau replied:
when you set a null output schema at configure time, it means the schema can either be variable or it is not known until runtime. If schema is null and format is not text, you should be setting output schema here to null.

albertshau commented on core-plugins/src/main/java/co/cask/hydrator/plugin/batch/source/PathTrackingInputFormat.java:
```diff
@@ -98,7 +98,10 @@ public static Schema getTextOutputSchema(@Nullable String pathField) {
     String path = filenameOnly ? fileSplit.getPath().getName() : fileSplit.getPath().toUri().toString();
     String schema = context.getConfiguration().get(SCHEMA);
-    Schema parsedSchema = Schema.parseJson(schema);
+    Schema parsedSchema = null;
```
nit: `Schema parsedSchema = schema == null ? null : Schema.parseJson(schema)`

albertshau commented on core-plugins/src/main/java/co/cask/hydrator/plugin/batch/source/PathTrackingInputFormat.java:
```diff
@@ -140,13 +143,19 @@ public NullWritable getCurrentKey() throws IOException, InterruptedException {
     protected abstract StructuredRecord getCurrentValue(StructuredRecord.Builder recordBuilder)
       throws IOException, InterruptedException;
 
+    protected abstract StructuredRecord getCurrentValueNull() throws IOException, InterruptedException;
```
this is a strange contract. The fact that you return null from one of the implementations and expect it never to be used by the parent means it's not a clean abstraction.

albertshau replied:
instead of this, reverse the order and change getCurrentValue to `StructuredRecord.Builder startCurrentValue()`. This class calls `startCurrentValue`, then sets the path field for it.

albertshau commented on core-plugins/src/main/java/co/cask/hydrator/plugin/common/AvroToStructuredTransformer.java:
```diff
@@ -71,4 +74,26 @@ public StructuredRecord transform(StructuredRecord.Builder builder, GenericRecor
     return builder.build();
   }
+
+  public StructuredRecord transform(GenericRecord genericRecord, String addFieldKey, String addFieldValue)
```
This isn't a very natural place for this logic, as this class should be about converting records. Modify the other transform method to return a StructuredRecord.Builder and remove the builder as an argument. The logic to add a field to the schema should be in the input format class, not here.

albertshau commented on core-plugins/src/main/java/co/cask/hydrator/plugin/common/AvroToStructuredTransformer.java:
```diff
@@ -71,4 +74,26 @@ public StructuredRecord transform(StructuredRecord.Builder builder, GenericRecor
+    newFields.add(newField);
+    Schema newSchema = Schema.recordOf(schema.getRecordName(), newFields);
+
+    GenericRecordBuilder genericRecordBuilder = new GenericRecordBuilder(new org.apache.avro.Schema.Parser().
```
don't build a brand new record and convert it, you'll double the work. You should be converting the GenericRecord once, and setting the pathfield at the very end on the StructuredRecord.Builder.

albertshau commented on core-plugins/src/test/java/co/cask/hydrator/plugin/batch/source/FileBatchSourceTest.java:
```diff
@@ -585,13 +591,13 @@ public void testFileBatchInputFormatParquet() throws Exception {
     DataSetManager<Table> outputManager = getDataset(outputDatasetName);
     List<StructuredRecord> output = MockSink.readOutput(outputManager);
-    Assert.assertEquals(expected, output);
+    Assert.assertEquals(output, expected);
```
expected is always first.

albertshau commented on core-plugins/src/test/java/co/cask/hydrator/plugin/batch/source/FileBatchSourceTest.java:
```diff
@@ -686,18 +772,33 @@ private ApplicationManager createSourceAndDeployApp(String appName, File file, S
-                                .build(),
-                              null));
+    ETLStage source;
+    if (schema != null) {
```
this is unnecessary repeat code. The only thing you need to do if schema is not null is put the schema property in the ImmutableMap builder

albertshau commented on core-plugins/src/main/java/co/cask/hydrator/plugin/batch/source/AbstractFileBatchSource.java:
```diff
@@ -99,7 +99,11 @@ public void configurePipeline(PipelineConfigurer pipelineConfigurer) {
       pipelineConfigurer.getStageConfigurer().setOutputSchema(config.getSchema());
     } else {
-      pipelineConfigurer.getStageConfigurer().setOutputSchema(DEFAULT_SCHEMA);
+      if (config.format.equalsIgnoreCase("text")) {
```
this is still not correct. you need to set it to the output schema if you know it. If schema is set provided, you should have validated it already and you can set it as the output schema. If schema is not provided but format is text, you can determine the correct output schema depending on if pathfield is set. This is exactly what you're doing elsewhere in the code, when setting schema in the Configuration. If schema is not provided and format is not text, you set output schema to null.

albertshau commented on core-plugins/src/main/java/co/cask/hydrator/plugin/batch/source/PathTrackingInputFormat.java:
```diff
@@ -83,6 +83,16 @@ public static Schema getTextOutputSchema(@Nullable String pathField) {
 
+  public static Schema addFieldToSchema(Schema schema, String addFieldKey) {
+    List<Schema.Field> newFields = new ArrayList<>();
+    for (Schema.Field field : schema.getFields()) {
```
replace this with
```
newFields = new ArrayList<>(schema.getFields().size() + 1);
newField.addAll(schema.getFields());
```

albertshau commented on core-plugins/src/main/java/co/cask/hydrator/plugin/batch/source/PathTrackingInputFormat.java:
```diff
@@ -212,13 +208,13 @@ public void initialize(InputSplit split, TaskAttemptContext context) throws IOEx
+        return recordTransformer.transform(delegate.getCurrentKey().datum(), schema, pathField);
+      }
+      Schema schema = recordTransformer.generateStructuredSchema(delegate.getCurrentKey().datum());
+      return recordTransformer.transform(delegate.getCurrentKey().datum(), addFieldToSchema(schema, pathField),
```
this will break when path field is null, you should only add the field if path is not null.

albertshau replied:
should do something like:
```
GenericRecord genericRecord = delegate.getCurrentKey().datum();
Schema recordSchema = schema;
if (schema == null) {
  recordSchema = recordTransformer.generateStructuredSchema(genericRecord.getSchema());
  if (pathField != null) {
    recordSchema = addFieldToSchema(recordSchema, pathField);
  }
}
return recordTransformer.transform(genericRecord, recordSchema, pathField);

```

albertshau commented on core-plugins/src/main/java/co/cask/hydrator/plugin/common/AvroToStructuredTransformer.java:
```diff
@@ -58,42 +43,34 @@ public StructuredRecord transform(GenericRecord genericRecord, Schema structured
 
-  public StructuredRecord transform(StructuredRecord.Builder builder, GenericRecord genericRecord,
-                                    Schema structuredSchema, String skipField)
+  public StructuredRecord.Builder transform(GenericRecord genericRecord, Schema structuredSchema, String skipField)
```
should annotate skipField as `@Nullable`

albertshau commented on core-plugins/src/main/java/co/cask/hydrator/plugin/common/AvroToStructuredTransformer.java:
```diff
@@ -58,42 +43,34 @@ public StructuredRecord transform(GenericRecord genericRecord, Schema structured
-    }
-    newFields.add(newField);
-    Schema newSchema = Schema.recordOf(schema.getRecordName(), newFields);
+  public Schema generateStructuredSchema(GenericRecord genericRecord) throws IOException {
```
take an avro Schema as the argument instead of a Generic Record and rename it convertSchema

sheivin commented:
@albertshau addressed comments. I had forgotten about the point related to the first comment earlier that the default schema (when schema is not set) can still have the pathField if it is not null (in the case of text). Now I am calling the function from PathTrackingInputFormat to set the schema if the format is text.

albertshau commented on core-plugins/src/main/java/co/cask/hydrator/plugin/batch/source/AbstractFileBatchSource.java:
```diff
@@ -96,7 +96,16 @@ public void configurePipeline(PipelineConfigurer pipelineConfigurer) {
+    if (config.getSchema() != null) {
+      pipelineConfigurer.getStageConfigurer().setOutputSchema(config.getSchema());
+    } else {
+      if (config.format.equalsIgnoreCase("text")) {
```
you can have one layer of ifs instead of having a nested if.

```
if (config.getSchema() != null) {
...
} else if ("text".equalsIgnoreCase(config.format)) {
...
} else {
...
}

albertshau commented on core-plugins/src/main/java/co/cask/hydrator/plugin/common/AvroToStructuredTransformer.java:
```diff
@@ -55,20 +45,32 @@ public StructuredRecord transform(GenericRecord genericRecord, Schema structured
-                                    Schema structuredSchema, String skipField)
-    throws IOException {
+  public StructuredRecord.Builder transform(GenericRecord genericRecord, Schema structuredSchema, @Nullable String
+    skipField) throws IOException {
```
variable type and name should be on the same line.

albertshau reviewed (commented):
a couple style comments, otherwise lgtm. fix and merge.
